File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (!root) return 0;
        return 1 + max_depth(root);
    }
    int max_depth(TreeNode*& root){
        if (!root) return 0;
        int leftdepth = 0;
        int rightdepth = 0;
        if (root->left){
            leftdepth = 1 + max_depth(root->left);
        }
        if (root->right){
            rightdepth = 1 + max_depth(root->right);
        }

        if (leftdepth > rightdepth) return leftdepth;
        return rightdepth;


    }
};

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root)->str:
        st = []
        def dfs(node):
            if node:
                st.append(f"{node.val}")
                dfs(node.left)
                dfs(node.right)
            else:
                st.append(f"n")
        dfs(root)
        return ",".join(st)
        
    # Decodes your encoded data to tree.
    def deserialize(self, data):
        data = data.split(",")
        it = iter(data)

        def dfs():
            tok = next(it)
            if tok == "n":
                return None

            node = TreeNode(int(tok))
            node.left = dfs()
            node.right = dfs()
            return node
        return dfs()
